fix: Tighten fuzz bounds with vm.assume((r & 3) < 2)

MITHRADEVI-K opened the pull request:
Reopening continuation of PR #133 (previously closed due to accidental fork deletion).

**Issue Addressed:** Improve fuzz testing CI (#67)  
**Changes Made:**
- Tightened fuzz bounds in `ExecuteBatch.t.sol` with `vm.assume((r & 3) < 2)`
- Reused `setUp()` accounts to reduce gas snapshot noise.

**Note:**  
The original PR was closed because the head repository was deleted.  
This PR contains the exact same commits (commit: `a31d7b3`) restored from local.

@cb-heimdall @wilsoncusack  ,please re-review when possible

MITHRADEVI-K committed:
fix: Tighten fuzz bounds with vm.assume((r & 3) < 2)

Closes #67

cb-heimdall commented:
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>
<!-- CODEOWNERS & Bar Raisers -->
<!-- Power Rangers -->
<!-- Links -->

MITHRADEVI-K reviewed (commented):
## Suggested Additional Changes

### 1. Add Deterministic Test for Edge Cases in ExecuteBatch.t.sol

The new `vm.assume((r & 3) < 2)` constraint excludes the edge cases where n=2 and n=3 MockTarget contracts are deployed. These batch sizes were previously tested via fuzzing but are now out of scope. To maintain coverage for these important edge cases, add a new deterministic test function after the existing `testExecuteBatch(uint256 r)` function:

```solidity
/// @dev Deterministic test covering edge cases n=2 and n=3 excluded from fuzz bounds.
/// The vm.assume((r & 3) < 2) constraint limits fuzzing to n=0,1 to reduce gas variance.
/// This test explicitly validates batch execution with larger payloads (n=2,3) to ensure
/// the bounds don't inadvertently skip critical integration scenarios.
function testExecuteBatchDeterministicEdgeCases() public {
    for (uint256 n = 2; n <= 3; ++n) {
        vm.prank(signer);
        account.addOwnerAddress(address(this));
        vm.deal(address(account), 10 ether);

        CoinbaseSmartWallet.Call[] memory calls = new CoinbaseSmartWallet.Call[](n);

        unchecked {
            for (uint256 i; i != n; ++i) {
                uint256 v = (i + 1) * 100; // Deterministic value for predictability
                calls[i].target = address(new MockTarget());
                calls[i].value = v;
                calls[i].data = abi.encodeWithSignature("setData(bytes)", _randomBytes(v));
            }
        }

        account.executeBatch(calls);

        // Verify all calls executed successfully
        for (uint256 i; i != n; ++i) {
            assertEq(calls[i].target.balance, (i + 1) * 100);
        }
    }
}
```

**Rationale:** Removing fuzzing constraints for n=2,3 is appropriate for reducing gas variance, but these values remain important test cases representing realistic multi-call batch scenarios.

### 2. Add Nonce Assumption in CoinbaseSmartWalletFactory.t.sol

In the fuzz test containing the `nonce` parameter, add an assumption to prevent pathologically large values. This is likely in a function like `testProxyDeployments(uint256 nonce)` or similar:

```solidity
/// @dev Bound nonce parameter to avoid excessive memory/computation during fuzzing.
/// Pathologically large nonce values can create unrealistic test scenarios with negligible
/// real-world probability. This assumption focuses fuzzing on practical nonce ranges.
vm.assume(nonce < type(uint128).max);
```

**Rationale:** Similar to the r bounds in ExecuteBatch, nonce values benefit from being bounded. This prevents the fuzzer from spending effort on unrealistic scenarios (nonce values > uint128) that are unlikely to occur in practice.

---

MITHRADEVI-K committed:
test: add deterministic batch execution edge case test for n=2,3

Add testExecuteBatchDeterministicEdgeCases() function to ExecuteBatch.t.sol that explicitly tests batch execution with n=2 and n=3 MockTarget contracts. This complements the fuzz test and ensures edge cases are properly covered with deterministic logic.

MITHRADEVI-K committed:
test: add nonce fuzz bounds to CoinbaseSmartWalletFactory tests

Add vm.assume(nonce < type(uint128).max) to test_createAccount_emitsAccountCreatedEvent() fuzz test in CoinbaseSmartWalletFactory.t.sol to prevent excessive gas costs and overflow issues during fuzzing, as the factory's internal nonce tracking uses uint128.

MITHRADEVI-K commented:
@amiecorso 

Would you be able to review this PR when you have a moment?

This is a continuation of #133. The PR tightens fuzz bounds and adds deterministic tests to improve fuzz testing CI as outlined in issue #67.

All changes are complete and ready for review. Any concerns or feedback from your side would help me refine this further.

Thank you.

MITHRADEVI-K commented:
Hi @amiecorso @wilsoncusack @cb-heimdall @zerggg @mashaw 

Gentle ping 

This PR (#134) is ready for final review:
- Tightens fuzz bounds to fix flaky CI (issue #67)
- Adds deterministic tests for n=2,3 edge cases that were excluded by the assume
- Adds reasonable nonce bounds in factory tests
- All tests passing, no gas regressions

It’s a continuation of previously discussed #133 and just needs one approving review from a maintainer to merge.

Would really appreciate it if anyone gets a chance to take a quick look — happy to hop on a 5-min call or make any tweaks instantly.

Thank you!  
Mithra
